plan/plans/join.go

```diff
@@ -342,3 +347,167 @@ func appendRow(prefix []interface{}, in []interface{}) []interface{} {
 	row = append(row, in...)
 	return row
 }
+
+// Next implements plan.Plan Next interface.
+func (r *JoinPlan) Next(ctx context.Context) (row *plan.Row, err error) {
+	if r.Right == nil {
+		return r.Left.Next(ctx)
+	}
+	if r.evalArgs == nil {
+		r.evalArgs = map[interface{}]interface{}{}
+	}
+	switch r.Type {
+	case LeftJoin:
+		return r.nextLeftJoin(ctx)
+	case RightJoin:
+		return r.nextRightJoin(ctx)
+	default:
+		return r.nextCrossJoin(ctx)
+	}
+}
+
+func (r *JoinPlan) nextLeftJoin(ctx context.Context) (row *plan.Row, err error) {
+	for {
+		if r.cursor < len(r.matchedRows) {
+			row = r.matchedRows[r.cursor]
+			r.cursor++
+			return
+		}
+		r.cursor = 0
+		var leftRow *plan.Row
+		leftRow, err = r.Left.Next(ctx)
+		if leftRow == nil || err != nil {
+			return nil, errors.Trace(err)
+		}
+		err = r.findMatchedRows(ctx, leftRow, false)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+}
+
+func (r *JoinPlan) nextRightJoin(ctx context.Context) (row *plan.Row, err error) {
+	for {
+		if r.cursor < len(r.matchedRows) {
+			row = r.matchedRows[r.cursor]
+			r.cursor++
+			return
+		}
+		var rightRow *plan.Row
+		rightRow, err = r.Right.Next(ctx)
+		if rightRow == nil || err != nil {
+			return nil, errors.Trace(err)
+		}
+		err = r.findMatchedRows(ctx, rightRow, true)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+}
+
+func (r *JoinPlan) findMatchedRows(ctx context.Context, row *plan.Row, right bool) (err error) {
+	var p plan.Plan
+	if right {
+		p = r.Left
+	} else {
+		p = r.Right
+	}
+	r.cursor = 0
+	r.matchedRows = nil
+	p.Close()
+	for {
+		var cmpRow *plan.Row
+		cmpRow, err = p.Next(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if cmpRow == nil {
+			break
+		}
+		var joined []interface{}
+		if right {
+			joined = append(cmpRow.Data, row.Data...)
+		} else {
+			joined = append(row.Data, cmpRow.Data...)
+		}
+		r.evalArgs[expressions.ExprEvalIdentFunc] = func(name string) (interface{}, error) {
+			return getIdentValue(name, r.Fields, joined, field.DefaultFieldFlag)
+		}
+		var b bool
+		b, err = expressions.EvalBoolExpr(ctx, r.On, r.evalArgs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if b {
+			cmpRow.Data = joined
+			cmpRow.RowKeys = append(row.RowKeys, cmpRow.RowKeys...)
+			r.matchedRows = append(r.matchedRows, cmpRow)
+		}
+	}
+	if len(r.matchedRows) == 0 {
+		if right {
+			leftLen := len(r.Fields) - len(r.Right.GetFields())
+			row.Data = append(make([]interface{}, leftLen), row.Data...)
+		} else {
+			rightLen := len(r.Fields) - len(r.Left.GetFields())
+			row.Data = append(row.Data, make([]interface{}, rightLen)...)
+		}
+		r.matchedRows = append(r.matchedRows, row)
+	}
+	return nil
+}
+
+func (r *JoinPlan) nextCrossJoin(ctx context.Context) (row *plan.Row, err error) {
+	for {
+		if r.curRow == nil {
```

Comment:
https://dev.mysql.com/doc/refman/5.7/en/nested-loop-joins.html

for the common join algorithm, we should first do left Next, and travel right iteratively.